File: cdm/java/com/kitware/physiology/datamodel/patient/conditions/SEPulmonaryFibrosis.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package com.kitware.physiology.datamodel.patient.conditions;

import com.kitware.physiology.cdm.PatientConditions.PulmonaryFibrosisData;

import com.kitware.physiology.datamodel.properties.SEScalar0To1;

public class SEPulmonaryFibrosis extends SEPatientCondition
{
  protected SEScalar0To1 severity;
  
  public SEPulmonaryFibrosis()
  {
    severity = null;
  }
  
  public void reset()
  {
    super.reset();
    if (severity != null)
      severity.invalidate();
  }
  
  public void copy(SEPulmonaryFibrosis other)
  {
    if(this==other)
      return;
    super.copy(other);
    if (other.severity != null)
      getSeverity().set(other.getSeverity());
    else if (severity != null)
      severity.invalidate();
  }
  
  public static void load(PulmonaryFibrosisData src, SEPulmonaryFibrosis dst) 
  {
    SEPatientCondition.load(src.getPatientCondition(), dst);
    if(src.hasSeverity())
      SEScalar0To1.load(src.getSeverity(),dst.getSeverity());
  }
  
  public static PulmonaryFibrosisData unload(SEPulmonaryFibrosis src) 
  {
    PulmonaryFibrosisData.Builder dst = PulmonaryFibrosisData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  
  protected static void unload(SEPulmonaryFibrosis src, PulmonaryFibrosisData.Builder dst)
  {
    SEPatientCondition.unload(src, dst.getPatientConditionBuilder());
    if (src.hasSeverity())
      dst.setSeverity(SEScalar0To1.unload(src.severity));
  }
  
  public boolean hasSeverity()
  {
    return severity == null ? false : severity.isValid();
  }
  public SEScalar0To1 getSeverity()
  {
    if (severity == null)
      severity = new SEScalar0To1();
    return severity;
  }
  
  public String toString()
  {
    return "Pulmonary Fibrosis" 
        + "\n\tSeverity: " + getSeverity();
  }

  
}

File: src/cpp/cdm/patient/conditions/SEPulmonaryFibrosis.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "cdm/CommonDefs.h"
#include "cdm/patient/conditions/SEPulmonaryFibrosis.h"
#include "cdm/properties/SEScalar0To1.h"
#include "cdm/io/protobuf/PBPatientConditions.h"

SEPulmonaryFibrosis::SEPulmonaryFibrosis(Logger* logger) : SEPatientCondition(logger)
{
  m_Severity = nullptr;
}

SEPulmonaryFibrosis::~SEPulmonaryFibrosis()
{
  SAFE_DELETE(m_Severity);
}

void SEPulmonaryFibrosis::Clear()
{
  SEPatientCondition::Clear();
  INVALIDATE_PROPERTY(m_Severity);
}

void SEPulmonaryFibrosis::Copy(const SEPulmonaryFibrosis& src)
{
  PBPatientCondition::Copy(src, *this);
}

bool SEPulmonaryFibrosis::IsValid() const
{
  return HasSeverity();
}

bool SEPulmonaryFibrosis::IsActive() const
{
  if (!IsValid())
    return false;
  return GetSeverity()>0;
}

bool SEPulmonaryFibrosis::HasSeverity() const
{
  return m_Severity == nullptr ? false : m_Severity->IsValid();
}
SEScalar0To1& SEPulmonaryFibrosis::GetSeverity()
{
  if (m_Severity == nullptr)
    m_Severity = new SEScalar0To1();
  return *m_Severity;
}
double SEPulmonaryFibrosis::GetSeverity() const
{
  if (m_Severity == nullptr)
    return SEScalar::dNaN();
  return m_Severity->GetValue();
}

void SEPulmonaryFibrosis::ToString(std::ostream &str) const
{
  SEPatientCondition::ToString(str);
  str << "\n\tSeverity: "; HasSeverity() ? str << *m_Severity : str << "NaN";
  str << std::flush;
}
